text/delegatable_cdd/README.md

```diff
@@ -0,0 +1,499 @@
+unlike some other schemes where the messages are elements of finite fields. The
+verification key will be in a different group than the message so if the message
+is in G1, verification key would be in G2 and vice-versa. Now we saw above that
+the delegatee's verkey is signed as part of the delegated credential. So if the
+delegator has verkey in group G2, it will sign messages in group G1 which means
+the delegatee needs to have a verkey in group G1. Similarly when this delegatee
+issues a credential, it will sign the subsequent delegatee's verkey which needs
+to be in G2 and these alternations will continue. Two variations of Groth
+signatures (described in the paper) are used, the algorithms are same but one,
+called `Groth1`, has messages in G1 and verkey in G2 and the other, called
+`Groth2`, has messages in G2 and verkey in G1. Even level issuer use Groth1
+signature and have verkey in group G2 while odd level issuers use Groth2
+signatures and have verkey in G1. Signing key is a single field element whereas
+the verification key size depends linearly on the maximum number of supported
+attributes.
+
+## API
+
+### Groth1 and Groth2 signatures
+
+- First, some public system parameters need to be generated which are used by
+  all system participants for generating keys, signing, verification, etc. These
+  parameters can be generated by calling `GrothS1::setup` or `GrothS2::setup`.
+  `setup` takes the maximum number of attributes that need to be supported. Keep
+  it one more than the number you want to support to accommodate the public key.
+    ```rust
+    let max_attrs = 10;
+    let label = "some label for generating NUMS".as_bytes();
+    let params1: Groth1SetupParams = GrothS1::setup(max_attrs, label);
+    // OR for Groth2
+    let params2: Groth2SetupParams = GrothS2::setup(max_attrs, label);
+    ```
+- Signing keys can be generated by calling `GrothS1::keygen` or `GrothS2::keygen`. Takes the corresponding setup parameters.
+    ```rust
+    let (sk1: GrothSigkey, vk1: Groth1Verkey) = GrothS1::keygen(&params1);
+    // OR for Groth2
+    let (sk2: GrothSigkey, vk2: Groth2Verkey) = GrothS2::keygen(&params2);
+    ```
+- A new signature can be created by calling `Groth1Sig:new` or `Groth2Sig:new`. 
+    ```rust
+    // `msgs` is a collection of elements in G1 
+    let sig: Groth1Sig = Groth1Sig::new(msgs.as_slice(), &sk1, &params1)?;
+    // OR for Groth2
+    let sig: Groth2Sig = Groth2Sig::new(msgs.as_slice(), &sk2, &params2)?;
+    ```
+- An existing signature can be randomized by calling `randomize` on the signature with a randomly chosen field element.
+    ```rust
+    let r = FieldElement::random();
+    let sig_randomized = sig.randomize(&r);
+    ```
+- 2 methods for signature verification, `verify` and `verify_fast`, both with the same API. `verify` computes several pairings to verify the signature whereas `verify_fast` does only 1 big multi-pairing.
```

Comment:
Because `verify_fast` has a negligibly higher probability of failure than `verify`. It's `O(n)/|F|` where `n` is the number of attributes and `|F|` is the size of finite field. I was not sure whether people will still sometime want `verify`.